### PR TITLE
Convert chart models to TS and export additional modules

### PR DESCRIPTION
**Pull request by kristw**
🏆 Enhancements
- Export additional modules from `@superset-ui/chart/query`

🏠 Internal
- Convert several models in `@superset-ui/chart` to TypeScript

### CODE REVIEW

File: packages/superset-ui-chart/src/@types/index.d.ts [hunk 1]
```diff
@@ -0,0 +1 @@
+declare module '@superset-ui/core';
```
Comment (williaster) [inline packages/superset-ui-chart/src/@types/index.d.ts]: what's the reasoning with adding the `@` prefix to the index file? is that best practice/standard? (tried to google for it but didn't see anything, a bit hard to search for). in `connection we have a `types.ts` file without an `@`.

File: packages/superset-ui-chart/src/models/ChartMetadata.ts [hunk 2]
```diff
@@ -0,0 +1,55 @@
+interface LookupTable {
+  [key: string]: boolean;
+}
+
+interface ChartMetadataConfig {
```
Comment (williaster) [inline packages/superset-ui-chart/src/models/ChartMetadata.ts]: should we export the interface for consumers?

File: packages/superset-ui-chart/src/@types/index.d.ts (hunk 1 above)
Comment (kristw) [reply packages/superset-ui-chart/src/@types/index.d.ts]: This is to manually declare types and fix the error for external module that do not have types. 

```
error TS7016: Could not find a declaration file for module '@superset-ui/core'
```

https://medium.com/@chris_72272/migrating-to-typescript-write-a-declaration-file-for-a-third-party-npm-module-b1f75808ed2

File: packages/superset-ui-chart/src/models/ChartMetadata.ts (hunk 2 above)
Comment (kristw) [reply packages/superset-ui-chart/src/models/ChartMetadata.ts]: ah right.

File: packages/superset-ui-chart/src/@types/index.d.ts (hunk 1 above)
Comment (kristw) [reply packages/superset-ui-chart/src/@types/index.d.ts]: Seems like it works if I just move this file to `src` with any name but extension `.d.ts` as well.

File: packages/superset-ui-chart/src/models/ChartPlugin.ts [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+      metadata,
+      buildQuery,
+      loadBuildQuery,
+      transformProps = IDENTITY,
+      loadTransformProps,
+      Chart,
+      loadChart,
+    } = config;
+    this.metadata = metadata;
+    this.loadBuildQuery = loadBuildQuery || (buildQuery ? () => buildQuery : null);
```
Comment (williaster) [inline packages/superset-ui-chart/src/models/ChartPlugin.ts]: not sure if we should be consistent with `null`s vs `undefined`. previously I was a fan of `null`s but our TS style guide says

> Use undefined in general
>   *Why?* Compatible with optionals (the ? operator). Triggers default argument initializers. 
> Use null where its a part of the API or conventional
>  e.g., optional wouldn't be appropriate if you know a server will respond with a string or null / won't be omitted

Comment (kristw) [reply packages/superset-ui-chart/src/models/ChartPlugin.ts]: Good point. @conglei added this part but I think either `undefined` or `null` would work for him. Will change to `undefined`

File: packages/superset-ui-chart/src/query/Metric.ts [hunk 4]
```diff
@@ -1,5 +1,5 @@
 import { Column } from './Column';
-import FormData from './FormData';
+import { FormData } from './FormData';
 
 export const LABEL_MAX_LENGTH = 43;
```
Comment (xtinec) [inline packages/superset-ui-chart/src/query/Metric.ts]: Why `.filter(metric => metric)` at line 60? (sorry github won't let me give a comment on that line).

Also, I found a bug in my original implication when putting the demo yesterday (metrics in form_data can appear as arrays of metrics). I can put up a fix against this branch.

Comment (kristw) [reply packages/superset-ui-chart/src/query/Metric.ts]: Filter the undefined field

**Pull request merged**